Fix race in CUDA initialization

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25788 Fix race in CUDA initialization**

Previously, I thought that _lazy_init held the GIL throughout initialization, so
I could write the code in a single-threaded manner.  This is not true; it
releases the GIL at various points, which make it possible for another thread to
race with initialization.

The correct fix is to add locking for the initialization section, so other
threads wait until the first thread finishes initializing before being let
in.  There is some subtlety with how to handle lazy calls, which will call
_lazy_init reentrantly; this is handled using TLS that lets you know if you
are the initializing thread (and therefore reentrant calls are OK.)

Fixes #16559

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17366348](https://our.internmc.facebook.com/intern/diff/D17366348)

ezyang committed:
Fix race in CUDA initialization

Previously, I thought that _lazy_init held the GIL throughout initialization, so
I could write the code in a single-threaded manner.  This is not true; it
releases the GIL at various points, which make it possible for another thread to
race with initialization.

The correct fix is to add locking for the initialization section, so other
threads wait until the first thread finishes initializing before being let
in.  There is some subtlety with how to handle lazy calls, which will call
_lazy_init reentrantly; this is handled using TLS that lets you know if you
are the initializing thread (and therefore reentrant calls are OK.)

Fixes #16559

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

soumith commented on torch/cuda/__init__.py:
```diff
@@ -181,16 +195,19 @@ def _lazy_init():
                    "CUDA call was originally invoked at:\n\n{}").format(str(e), orig_traceback)
             raise_from(DeferredCudaCallError(msg), e)
+    _initialized = True
+    _initialization_lock.release()
```
if dequeing the `_queued_calls` raises an exception for whatever reason, you never get to the `.release()` line, does this have bad implications?

ezyang replied:
All other threads / cuda calls will deadlock on `_lazy_init`. I was shocked that threading.Lock didn't have a context manager version...

ezyang committed:
Rebase, use contxt manager on "Fix race in CUDA initialization"


Previously, I thought that _lazy_init held the GIL throughout initialization, so
I could write the code in a single-threaded manner.  This is not true; it
releases the GIL at various points, which make it possible for another thread to
race with initialization.

The correct fix is to add locking for the initialization section, so other
threads wait until the first thread finishes initializing before being let
in.  There is some subtlety with how to handle lazy calls, which will call
_lazy_init reentrantly; this is handled using TLS that lets you know if you
are the initializing thread (and therefore reentrant calls are OK.)

Fixes #16559

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
Test failure seems to be legit.

```
Sep 09 16:26:19 test_wrong_cuda_fork (__main__.TestMultiprocessing) ... FAIL
Sep 09 16:26:19 
Sep 09 16:26:19 ======================================================================
Sep 09 16:26:19 FAIL: test_wrong_cuda_fork (__main__.TestMultiprocessing)
Sep 09 16:26:19 ----------------------------------------------------------------------
Sep 09 16:26:19 Traceback (most recent call last):
Sep 09 16:26:19   File "test_multiprocessing.py", line 492, in test_wrong_cuda_fork
Sep 09 16:26:19     self.assertRegex(results[1].decode('ascii'), "Cannot re-initialize CUDA in forked subprocess.")
Sep 09 16:26:19 AssertionError: Regex didn't match: 'Cannot re-initialize CUDA in forked subprocess.' not found in 'THCudaCheck FAIL file=/var/lib/jenkins/workspace/torch/csrc/cuda/Module.cpp line=37 error=3 : initialization error\nProcess Process-1:\nTraceback (most recent call last):\n  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap\n    self.run()\n  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 93, in run\n    self._target(*self._args, **self._kwargs)\n  File "<string>", line 4, in run\n  File "/opt/conda/lib/python3.6/site-packages/torch/cuda/__init__.py", line 297, in set_device\n    torch._C._cuda_setDevice(device)\nRuntimeError: cuda runtime error (3) : initialization error at /var/lib/jenkins/workspace/torch/csrc/cuda/Module.cpp:37\nTHCudaCheck FAIL file=/var/lib/jenkins/workspace/torch/csrc/cuda/Module.cpp line=37 error=3 : initialization error\nProcess Process-2:\nTraceback (most recent call last):\n  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap\n    self.run()\n  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 93, in run\n    self._target(*self._args, **self._kwargs)\n  File "<string>", line 4, in run\n  File "/opt/conda/lib/python3.6/site-packages/torch/cuda/__init__.py", line 297, in set_device\n    torch._C._cuda_setDevice(device)\nRuntimeError: cuda runtime error (3) : initialization error at /var/lib/jenkins/workspace/torch/csrc/cuda/Module.cpp:37\n'
Sep 09 16:26:19 
Sep 09 16:26:19 ----------------------------------------------------------------------
```

ezyang committed:
Correctly unset tls when done too on "Fix race in CUDA initialization"


Previously, I thought that _lazy_init held the GIL throughout initialization, so
I could write the code in a single-threaded manner.  This is not true; it
releases the GIL at various points, which make it possible for another thread to
race with initialization.

The correct fix is to add locking for the initialization section, so other
threads wait until the first thread finishes initializing before being let
in.  There is some subtlety with how to handle lazy calls, which will call
_lazy_init reentrantly; this is handled using TLS that lets you know if you
are the initializing thread (and therefore reentrant calls are OK.)

Fixes #16559

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Fix race in CUDA initialization"


Previously, I thought that _lazy_init held the GIL throughout initialization, so
I could write the code in a single-threaded manner.  This is not true; it
releases the GIL at various points, which make it possible for another thread to
race with initialization.

The correct fix is to add locking for the initialization section, so other
threads wait until the first thread finishes initializing before being let
in.  There is some subtlety with how to handle lazy calls, which will call
_lazy_init reentrantly; this is handled using TLS that lets you know if you
are the initializing thread (and therefore reentrant calls are OK.)

Fixes #16559

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Fix race in CUDA initialization"


Previously, I thought that _lazy_init held the GIL throughout initialization, so
I could write the code in a single-threaded manner.  This is not true; it
releases the GIL at various points, which make it possible for another thread to
race with initialization.

The correct fix is to add locking for the initialization section, so other
threads wait until the first thread finishes initializing before being let
in.  There is some subtlety with how to handle lazy calls, which will call
_lazy_init reentrantly; this is handled using TLS that lets you know if you
are the initializing thread (and therefore reentrant calls are OK.)

Fixes #16559

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]
